tests/nnstreamer_edge/nnstreamer-edge-custom-test.c

```diff
@@ -0,0 +1,229 @@
+  custom_h->is_connected = 0;
+
+  return NNS_EDGE_ERROR_NONE;
+}
+
+static int
+nns_edge_custom_connect (void *priv)
+{
+  nns_edge_custom_test_s *custom_h;
+  nns_edge_data_h data_h;
+  gchar *raw_data;
+  if (!priv) {
+    nns_loge ("Invalid param, handle should not be null.");
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+  }
+  custom_h = (nns_edge_custom_test_s *) priv;
+  custom_h->is_connected = 1;
+
+  /* Push dummy buffers to launch GstBaseSRc */
+  nns_edge_data_create (&data_h);
+  raw_data = g_strdup ("Dummy data");
+  nns_edge_data_add (data_h, raw_data, strlen (raw_data) + 1, g_free);
+  nns_edge_event_invoke_callback (custom_h->event_cb, custom_h->user_data,
+      NNS_EDGE_EVENT_NEW_DATA_RECEIVED, data_h, sizeof (nns_edge_data_h), NULL);
+
+  return NNS_EDGE_ERROR_NONE;
+}
+
+static int
+nns_edge_custom_subscribe (void *priv)
+{
+  UNUSED (priv);
+  return NNS_EDGE_ERROR_NOT_SUPPORTED;
+}
+
+static int
+nns_edge_custom_is_connected (void *priv)
+{
+  nns_edge_custom_test_s *custom_h;
+  if (!priv) {
+    nns_loge ("Invalid param, handle should not be null.");
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+  }
+  custom_h = (nns_edge_custom_test_s *) priv;
+
+  if (custom_h->is_connected == 1)
+    return NNS_EDGE_ERROR_NONE;
+
+  return NNS_EDGE_ERROR_CONNECTION_FAILURE;
+}
+
+static int
+nns_edge_custom_set_event_cb (void *priv, nns_edge_event_cb cb, void *user_data)
+{
+  nns_edge_custom_test_s *custom_h;
+  if (!priv) {
+    nns_loge ("Invalid param, handle should not be null.");
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+  }
+  custom_h = (nns_edge_custom_test_s *) priv;
+
+  custom_h->event_cb = cb;
+  custom_h->user_data = user_data;
+
+  return NNS_EDGE_ERROR_NONE;
+}
+
+static int
+nns_edge_custom_send_data (void *priv, nns_edge_data_h data_h)
+{
+  if (!priv || !data_h) {
+    nns_loge ("Invalid param, handle or data should not be null.");
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+  }
+
+  return NNS_EDGE_ERROR_NONE;
+}
+
+static int
+nns_edge_custom_set_info (void *priv, const char *key, const char *value)
+{
+  nns_edge_custom_test_s *custom_h;
+  if (!priv || !key || !value) {
+    nns_loge ("Invalid param, handle, key or value should not be null.");
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+  }
+  custom_h = (nns_edge_custom_test_s *) priv;
+
+  if (strcasecmp (key, "PEER_ADDRESS") == 0) {
+    SAFE_FREE (custom_h->peer_address);
+    custom_h->peer_address = g_strdup (value);
+    return NNS_EDGE_ERROR_NONE;
+  }
+
+  return NNS_EDGE_ERROR_NONE;
+}
+
+static int
+nns_edge_custom_get_info (void *priv, const char *key, char **value)
+{
+  nns_edge_custom_test_s *custom_h;
+  if (!priv || !key || !value) {
+    nns_loge ("Invalid param, handle, key or value should not be null.");
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+  }
+  custom_h = (nns_edge_custom_test_s *) priv;
+
+  if (strcasecmp (key, "PEER_ADDRESS") == 0) {
+    *value = g_strdup (custom_h->peer_address);
```

Comment:
Is there no case where the value is null?
For cases where we call ```nns_edge_custom_get_info``` before calling ```nns_edge_custom_set_info``` :smile:

Reply:
It can be `NULL`.
If the arg of `g_strdup` is `NULL`, it will return `NULL`.